routing/routing_session.cpp

```diff
@@ -502,7 +503,19 @@ void RoutingSession::GenerateNotifications(std::vector<std::string> & notificati
   // Generate turns notifications.
   std::vector<turns::TurnItemDist> turns;
   if (m_route->GetNextTurns(turns))
-    m_turnNotificationsMgr.GenerateTurnNotifications(turns, notifications);
+  {
+    std::string nextStreet;
+
+    // only get nextStreet if TtsStreetNames is enabled
+    if (shouldAnnounceStreetNames)
+    {
+      RouteSegment::RoadNameInfo targetRoadNameInfo;
+      m_route->GetNextTurnStreetName(targetRoadNameInfo);
+      GetFullRoadName(targetRoadNameInfo, nextStreet);
```

Comment:
If I’m not mistaken, `GetNextTurnStreetName()` walks the routing graph to find the actual next street name:

https://github.com/organicmaps/organicmaps/blob/67079e48fb1d5dede4997d6f254c9cb83ef75773/routing/route.cpp#L175-L181

This is ideal behavior for the initial instruction, since users don’t necessarily intuit which way would be “heading west”, for example. It’s also great for roundabout instructions. However, it can be misleading in the context of a freeway/expressway exit instruction: the user doesn’t necessarily see a street name in advance on any signs. When they do see a street name, it may not be the same one that they need to take.

In the U.S., street names are very often signposted as freeway/expressway exit destinations, typically but not necessarily listed before other destinations. [There’s a long-running dispute](https://community.openstreetmap.org/t/usage-of-the-destination-and-destination-street-tag-with-the-street-name/101788) about whether to include them in the main `destination` tag or split it out into a separate `destination:street` tag. The latter approach forces routers to assume that the street is listed first and also assume that it isn’t listed in fine print as “alt text” for a route shield.

Reply:
@1ec5 if you're concerned that this code would result in behavior like telling a user to exit Highway 101 by saying "in 3000 feet, exit onto Main Street" in the US, don't worry. Some magic within this bit of code (maybe `GetClosestStreetNameAfterIdx()`?) grabs the appropriate tags from the highway exit nodes and so the behavior is to instead say "in 3000 feet, exit onto Exit 123; Main Street; Embarcadero" -- it essentially pronounces the same text that's been displayed at the top of the screen this whole time.

Obviously this can be improved further as discussed in the PR, but the information is there and it's a vast improvement over the status quo of being on a highway and it says "in 3000 feet, exit. ... ... ... exit." and you have to figure out whether they mean Exit 122, Exit 123A, or Exit 123B in the spaghetti streets typical to our experience.

There's a beta APK available in the PR, I encourage you to download and try it out!